File: core/stockitem.go
package core

import (
	"github.com/louisevanderlith/husk/hsk"
	"github.com/louisevanderlith/husk/keys"
	"github.com/louisevanderlith/husk/validation"
	"time"
)

type StockItem struct {
	ItemKey       *keys.TimeKey
	ShortName     string `hsk:"size(128)"`
	ImageKey      *keys.TimeKey
	OwnerKey      *keys.TimeKey //Hero
	Expires       time.Time
	Currency      string
	Price         float32
	EstimateValue float32
	Tags          []string
	Location      string `hsk:"size(128)"`
	OwnerHistory  map[time.Time]hsk.Key
	Views         int64 `hsk:"null"`
}

func (s StockItem) Valid() error {
	err := validation.Struct(s)

	if err != nil {
		return err
	}

	return PriceInBounds(s.Price, s.EstimateValue)
}

File: lib/bodies/stockitem.dart
import 'package:mango_ui/keys.dart';

class StockItem {
  final Key itemKey;
  final String shortname;
  final Key imageKey;
  final Key ownerKey;
  final DateTime expires;
  final String currency;
  final double price;
  final double estimate;
  final List<String> tags;
  final String location;
  final int views;
  final Map<DateTime, Key> history;

  StockItem(
      this.itemKey,
      this.shortname,
      this.imageKey,
      this.ownerKey,
      this.expires,
      this.currency,
      this.price,
      this.estimate,
      this.tags,
      this.location,
      this.views,
      this.history);

  Map<String, dynamic> toJson() {
    return {
      "ItemKey": itemKey,
      "ImageKey": imageKey,
      "OwnerKey": ownerKey,
      "Expires": expires,
      "Price": price,
      "Tags": tags,
      "Location": location,
      "Views": views,
      "OwnerHistory": history,
      "EstimateValue": estimate,
      "Currency": currency,
      "ShortName": shortname,
    };
  }
}
